imitation/model_env.py

```diff
@@ -0,0 +1,346 @@
+"""Finite-horizon discrete environments with known transition dynamics. These
+are handy when you want to perform exact maxent policy optimisation."""
+
+import abc
+
+import gym
+from gym import spaces
+import numpy as np
+
+
+class ModelBasedEnv(gym.Env, metaclass=abc.ABCMeta):
+    """ABC for tabular environments with known dynamics."""
+    def __init__(self):
+        self.cur_state = 0
+        self.actions_taken = 0
+        # we must cache action & observation spaces instead of reconstructing
+        # anew so that the random state of the action space (!!) is preserved
+        self._action_space = None
+        self._observation_space = None
+        self.seed()
+
+    @property
+    def action_space(self):
+        if not self._action_space:
+            self._action_space = spaces.Discrete(self.n_actions)
+        return self._action_space
+
+    @property
+    def observation_space(self):
+        if not self._observation_space:
+            self._observation_space = spaces.Box(low=float('-inf'),
+                                                 high=float('inf'),
+                                                 shape=(self.obs_dim, ))
+        return self._observation_space
+
+    def seed(self, seed=None):
+        if seed is None:
+            seed = np.random.randint(0, 1 << 31)
+        self.rand_state = np.random.RandomState(seed=seed)
+        return [seed]
+
+    def reset(self):
+        self.cur_state = 0
+        self.actions_taken = 0
+        # as in step(), we copy so that it can't be mutated in-place (updates
+        # will be reflected in self.observation_matrix!)
+        return self.observation_matrix[self.cur_state].copy()
+
+    def step(self, action):
+        old_state = self.cur_state
+        out_dist = self.transition_matrix[old_state, action]
+        choice_states = np.arange(self.n_states)
+        next_state = int(
+            self.rand_state.choice(choice_states, p=out_dist, size=()))
+        self.cur_state = next_state
+        self.actions_taken += 1
+        done = self.actions_taken >= self.horizon
+        reward = self.reward_matrix[old_state]
+        assert np.isscalar(reward), reward
+        # copy so that it can't be mutated in-place
+        obs = self.observation_matrix[next_state].copy()
```

Comment:
The point is to support different featurisations. e.g RandomMDP has options for both one-hot observations (the identity matrix) and observations that are initially sampled from a Gaussian when the MDP is constructed. This is important, because if you just stick with the one-hot featurisation then there's no point in using anything other than a linear model for your reward function or policy. All the example envs still include "perfect information" in the sense that the observation vectors for each state are unique.

Reply:
OK, this makes sense, but really needs documenting. Can you add it to the `observation_matrix` property?